File: lib/blog/receipt_printer_service.ex
defmodule Blog.ReceiptPrinterService do
  @moduledoc """
  Service that polls for pending receipt messages and sends them to the printer,
  including support for dithered images.
  """
  use GenServer
  require Logger
  
  alias Blog.ReceiptMessages
  alias ReceiptPrinter.NetworkClient
  alias ReceiptPrinter.ReceiptBuilder
  alias ReceiptPrinter.ImageProcessor
  
  @poll_interval 5_000  # 5 seconds
  @printer_host "127.0.0.1"
  @printer_port 9100
  
  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end
  
  @impl true
  def init(opts) do
    state = %{
      poll_interval: Keyword.get(opts, :poll_interval, @poll_interval),
      printer_host: Keyword.get(opts, :printer_host, @printer_host),
      printer_port: Keyword.get(opts, :printer_port, @printer_port),
      enabled: Keyword.get(opts, :enabled, true)
    }
    
    if state.enabled do
      schedule_poll(state.poll_interval)
      Logger.info("Receipt Printer Service started - polling every #{state.poll_interval}ms")
    else
      Logger.info("Receipt Printer Service started but disabled")
    end
    
    {:ok, state}
  end
  
  @impl true
  def handle_info(:poll, state) do
    if state.enabled do
      process_pending_messages(state)
      schedule_poll(state.poll_interval)
    end
    
    {:noreply, state}
  end
  
  defp schedule_poll(interval) do
    Process.send_after(self(), :poll, interval)
  end
  
  defp process_pending_messages(state) do
    messages = ReceiptMessages.list_pending_messages()
    
    if length(messages) > 0 do
      Logger.info("Processing #{length(messages)} pending messages")
      
      Enum.each(messages, fn message ->
        try do
          print_message(message, state)
          ReceiptMessages.mark_as_printed(message)
          Logger.info("Successfully printed message #{message.id}")
        rescue
          e ->
            Logger.error("Failed to print message #{message.id}: #{inspect(e)}")
            ReceiptMessages.mark_as_failed(message)
        end
      end)
    end
  end
  
  defp print_message(message, state) do
    printer_opts = [
      host: state.printer_host,
      port: state.printer_port
    ]
    
    # Build the receipt with proper formatting
    builder = ReceiptBuilder.new()
    |> ReceiptBuilder.init()
    |> ReceiptBuilder.align(:center)
    |> ReceiptBuilder.text("========================")
    |> ReceiptBuilder.line_feed()
    |> ReceiptBuilder.font_size(:double)
    |> ReceiptBuilder.text("NEW MESSAGE")
    |> ReceiptBuilder.font_size(:normal)
    |> ReceiptBuilder.line_feed()
    |> ReceiptBuilder.text("========================")
    |> ReceiptBuilder.line_feed(2)
    |> ReceiptBuilder.align(:left)
    |> ReceiptBuilder.text("From: #{message.sender_ip || "unknown"}")
    |> ReceiptBuilder.line_feed()
    |> ReceiptBuilder.text("Time: #{format_datetime(message.inserted_at)}")
    |> ReceiptBuilder.line_feed()
    |> ReceiptBuilder.text("--------------------------------")
    |> ReceiptBuilder.line_feed(2)
    
    # Add message content
    builder = builder
    |> ReceiptBuilder.bold(true)
    |> ReceiptBuilder.text("Message:")
    |> ReceiptBuilder.bold(false)
    |> ReceiptBuilder.line_feed()
    
    # Split content into lines and add indentation
    message.content
    |> String.split("\n")
    |> Enum.reduce(builder, fn line, acc ->
      acc
      |> ReceiptBuilder.text("  #{line}")
      |> ReceiptBuilder.line_feed()
    end)
    |> ReceiptBuilder.line_feed()
    
    # Check if there's an image to print
    builder = if message.image_url && message.image_url != "" do
      builder
      |> ReceiptBuilder.text("--------------------------------")
      |> ReceiptBuilder.line_feed()
      |> ReceiptBuilder.align(:center)
      |> ReceiptBuilder.text("[Image Attached]")
      |> ReceiptBuilder.line_feed()
      |> print_image_if_available(message.image_url, printer_opts)
      |> ReceiptBuilder.align(:left)
    else
      builder
    end
    
    # Footer
    builder = builder
    |> ReceiptBuilder.text("--------------------------------")
    |> ReceiptBuilder.line_feed(2)
    |> ReceiptBuilder.align(:center)
    |> ReceiptBuilder.text("Thank you for your message!")
    |> ReceiptBuilder.line_feed()
    |> ReceiptBuilder.text("ID: MSG#{String.pad_leading(to_string(message.id), 6, "0")}")
    |> ReceiptBuilder.line_feed(4)
    |> ReceiptBuilder.cut(:partial)
    
    # Send to printer
    NetworkClient.print_receipt(builder, printer_opts)
  end
  
  defp print_image_if_available(builder, image_url, printer_opts) do
    # For now, if image_url is a path or binary data, process it
    # In a real implementation, you'd download the image from URL
    case process_image_for_printing(image_url) do
      {:ok, bitmap_data} ->
        # Add the raw bitmap data directly to the builder
        # This requires extending ReceiptBuilder to handle raw data
        builder
        |> ReceiptBuilder.raw(bitmap_data)
        |> ReceiptBuilder.line_feed(2)
      {:error, reason} ->
        Logger.warning("Could not process image: #{reason}")
        builder
        |> ReceiptBuilder.text("[Image could not be printed]")
        |> ReceiptBuilder.line_feed()
    end
  end
  
  defp process_image_for_printing(image_url) when is_binary(image_url) do
    # Check if it's a file path or URL
    cond do
      File.exists?(image_url) ->
        ImageProcessor.process_image(image_url, width: 384)
      
      String.starts_with?(image_url, "http") ->
        # Download image first
        case download_image(image_url) do
          {:ok, image_binary} ->
            ImageProcessor.process_image_binary(image_binary, width: 384)
          error ->
            error
        end
      
      true ->
        {:error, "Invalid image source"}
    end
  end
  
  defp download_image(url) do
    case :httpc.request(:get, {String.to_charlist(url), []}, [], []) do
      {:ok, {{_, 200, _}, _headers, body}} ->
        {:ok, :erlang.list_to_binary(body)}
      {:ok, {{_, status, _}, _, _}} ->
        {:error, "HTTP #{status}"}
      {:error, reason} ->
        {:error, inspect(reason)}
    end
  end
  
  defp format_datetime(nil), do: "Unknown"
  defp format_datetime(datetime) do
    Calendar.strftime(datetime, "%Y-%m-%d %H:%M:%S")
  end
end

File: receipt_printer_service.py
#!/usr/bin/env python3
"""
Receipt Printer Polling Service
Polls the blog API for pending messages and prints them locally
"""

import json
import time
import socket
import requests
import logging
import argparse
import sys
from datetime import datetime
from typing import Optional, Dict, List
import os
from pathlib import Path

# Hardcoded API key for receipt printer authentication
RECEIPT_PRINTER_API_KEY = "67656cfac9eea273fed7a403088874506ab41b700056d01f660537e2be7316f4"

logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class ReceiptPrinterService:
    def __init__(self, api_url: str, auth_token: str, printer_host: str = "127.0.0.1", 
                 printer_port: int = 9100, poll_interval: int = 10):
        self.api_url = api_url.rstrip('/')
        self.auth_token = auth_token
        self.printer_host = printer_host
        self.printer_port = printer_port
        self.poll_interval = poll_interval
        self.session = requests.Session()
        self.session.headers.update({
            'X-Auth-Token': auth_token,
            'Content-Type': 'application/json'
        })
        
    def fetch_pending_messages(self) -> List[Dict]:
        """Fetch pending messages from the API"""
        try:
            response = self.session.get(f"{self.api_url}/api/receipt_messages/pending")
            response.raise_for_status()
            data = response.json()
            return data.get('messages', [])
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to fetch messages: {e}")
            return []
    
    def mark_message_printed(self, message_id: int) -> bool:
        """Mark a message as printed in the API"""
        try:
            response = self.session.post(
                f"{self.api_url}/api/receipt_messages/{message_id}/printed"
            )
            response.raise_for_status()
            return True
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to mark message {message_id} as printed: {e}")
            return False
    
    def mark_message_failed(self, message_id: int) -> bool:
        """Mark a message as failed in the API"""
        try:
            response = self.session.post(
                f"{self.api_url}/api/receipt_messages/{message_id}/failed"
            )
            response.raise_for_status()
            return True
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to mark message {message_id} as failed: {e}")
            return False
    
    def format_receipt(self, message: Dict) -> bytes:
        """Format a message as ESC/POS receipt data"""
        ESC = b'\x1b'
        GS = b'\x1d'
        
        # Initialize printer
        commands = ESC + b'@'  # Initialize
        
        # Header
        commands += ESC + b'a\x01'  # Center align
        commands += b'=' * 24 + b'\n'
        commands += ESC + b'!\x30'  # Double height/width
        commands += b'NEW MESSAGE\n'
        commands += ESC + b'!\x00'  # Normal size
        commands += b'=' * 24 + b'\n\n'
        
        # Metadata
        commands += ESC + b'a\x00'  # Left align
        sender_text = f"From: {message.get('sender_name') or message.get('sender_ip', 'Unknown')}\n"
        commands += sender_text.encode('utf-8')
        
        timestamp = message.get('created_at', '')
        if timestamp:
            try:
                dt = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
                time_text = f"Time: {dt.strftime('%Y-%m-%d %H:%M:%S UTC')}\n"
                commands += time_text.encode('utf-8')
            except:
                pass
        
        commands += b'-' * 48 + b'\n\n'
        
        # Message content
        commands += b'Message:\n'
        content = message.get('content', '')
        
        # Word wrap at 48 characters
        wrapped_lines = self.wrap_text(content, 44)
        for line in wrapped_lines:
            commands += b'  ' + line.encode('utf-8') + b'\n'
        
        commands += b'\n'
        
        # Image indicator
        if message.get('image_url'):
            commands += b'-' * 48 + b'\n'
            commands += ESC + b'a\x01'  # Center
            commands += b'[Image Attached]\n'
            commands += ESC + b'a\x00'  # Left
            commands += b'\n'
        
        # Footer
        commands += b'-' * 48 + b'\n\n'
        commands += ESC + b'a\x01'  # Center
        commands += b'Thank you for your message!\n\n'
        
        # Message ID
        message_id = f"ID: MSG{message.get('id', 0):06d}\n"
        commands += message_id.encode('utf-8')
        
        # Cut paper
        commands += b'\n' * 3
        commands += GS + b'V\x42\x00'  # Partial cut
        
        return commands
    
    def wrap_text(self, text: str, width: int) -> List[str]:
        """Word wrap text to specified width"""
        lines = []
        for paragraph in text.split('\n'):
            if not paragraph:
                lines.append('')
                continue
                
            words = paragraph.split()
            current_line = ''
            
            for word in words:
                if not current_line:
                    current_line = word
                elif len(current_line) + 1 + len(word) <= width:
                    current_line += ' ' + word
                else:
                    lines.append(current_line)
                    current_line = word
            
            if current_line:
                lines.append(current_line)
        
        return lines
    
    def print_to_printer(self, data: bytes) -> bool:
        """Send data to the printer via TCP"""
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(5)
                sock.connect((self.printer_host, self.printer_port))
                sock.sendall(data)
                logger.info(f"Successfully sent {len(data)} bytes to printer")
                return True
        except Exception as e:
            logger.error(f"Failed to print: {e}")
            return False
    
    def process_message(self, message: Dict) -> bool:
        """Process a single message"""
        message_id = message.get('id')
        logger.info(f"Processing message {message_id} from {message.get('sender_ip')}")
        
        # Format receipt
        receipt_data = self.format_receipt(message)
        
        # Save to file for debugging
        debug_dir = Path("printed_receipts")
        debug_dir.mkdir(exist_ok=True)
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        debug_file = debug_dir / f"receipt_{message_id}_{timestamp}.bin"
        with open(debug_file, 'wb') as f:
            f.write(receipt_data)
        logger.debug(f"Saved receipt to {debug_file}")
        
        # Print
        if self.print_to_printer(receipt_data):
            logger.info(f"Successfully printed message {message_id}")
            self.mark_message_printed(message_id)
            return True
        else:
            logger.error(f"Failed to print message {message_id}")
            self.mark_message_failed(message_id)
            return False
    
    def run(self):
        """Main polling loop"""
        logger.info(f"Starting receipt printer service")
        logger.info(f"API URL: {self.api_url}")
        logger.info(f"Printer: {self.printer_host}:{self.printer_port}")
        logger.info(f"Poll interval: {self.poll_interval} seconds")
        
        while True:
            try:
                messages = self.fetch_pending_messages()
                
                if messages:
                    logger.info(f"Found {len(messages)} pending message(s)")
                    for message in messages:
                        self.process_message(message)
                        time.sleep(1)  # Small delay between prints
                
                time.sleep(self.poll_interval)
                
            except KeyboardInterrupt:
                logger.info("Service stopped by user")
                break
            except Exception as e:
                logger.error(f"Unexpected error: {e}")
                time.sleep(self.poll_interval)


def main():
    parser = argparse.ArgumentParser(description='Receipt Printer Polling Service')
    parser.add_argument('--api-url', default='https://www.bobbby.online',
                        help='Blog API URL')
    parser.add_argument('--auth-token', required=False,
                        help='API authentication token')
    parser.add_argument('--printer-host', default='127.0.0.1',
                        help='Printer host/IP')
    parser.add_argument('--printer-port', type=int, default=9100,
                        help='Printer port')
    parser.add_argument('--poll-interval', type=int, default=10,
                        help='Polling interval in seconds')
    parser.add_argument('--debug', action='store_true',
                        help='Enable debug logging')
    parser.add_argument('--print-text', type=str,
                        help='Print text directly without using API')
    parser.add_argument('--api-key', type=str,
                        help='API key for authentication')
    
    args = parser.parse_args()
    
    if args.debug:
        logging.getLogger().setLevel(logging.DEBUG)
    
    # Direct print mode - print text and exit
    if args.print_text:
        # Verify API key for direct printing
        if args.api_key != RECEIPT_PRINTER_API_KEY:
            logger.error("Invalid API key for direct printing")
            sys.exit(1)
        
        service = ReceiptPrinterService(
            api_url=args.api_url,
            auth_token='dummy',  # Not needed for direct printing
            printer_host=args.printer_host,
            printer_port=args.printer_port,
            poll_interval=args.poll_interval
        )
        
        # Create a simple message structure for direct printing
        message = {
            'id': int(time.time()),  # Use integer for ID
            'content': args.print_text,
            'sender_name': 'Direct Print',
            'created_at': datetime.now().isoformat()
        }
        
        # Format and print the receipt
        receipt_data = service.format_receipt(message)
        if service.print_to_printer(receipt_data):
            logger.info("Successfully printed text")
            sys.exit(0)
        else:
            logger.error("Failed to print text")
            sys.exit(1)
    
    # Normal polling mode
    if not args.auth_token:
        parser.error("--auth-token is required when not using --print-text")
    
    service = ReceiptPrinterService(
        api_url=args.api_url,
        auth_token=args.auth_token,
        printer_host=args.printer_host,
        printer_port=args.printer_port,
        poll_interval=args.poll_interval
    )
    
    try:
        service.run()
    except Exception as e:
        logger.error(f"Service failed: {e}")
        sys.exit(1)


if __name__ == '__main__':
    main()